### ISSUE TITLE
Incorrect auto-correction for `Style/StringLiterals` cop

### ISSUE THREAD

**Issue by konieczkow** [rubocop/rubocop#11123]
Hi there, thanks to all the people working on Rubocop. 

I noticed a weird behaviour when applying an autocorrection to the following code.

```ruby
a = {
  b: "{\"[\\\"*\\\"]\""
}
```

This gets correctly marked as violating `Style/StringLiterals` but when I run rubocop with `-a` (which I assume means it will correct all the cops as long as it is safe to do so) I get this as the output:

```ruby
a = {
  b: '{"["*"]"'
}

```

It seems to me that the autocorrection is too greedy and swollows too many escape characters?

--------

## Expected behavior

I would expect rubocop to either correctly escape the code to:
```ruby
a = {
  b: {"[\"*\"]"
}
```
or it should say that the cop is unsafe and can be used with `-A`, instead of `-a`.

## Actual behavior

Autocorrecting:

```ruby
a = {
  b: "{\"[\\\"*\\\"]\""
}
```

produces:

```ruby
a = {
  b: '{"["*"]"'
}

```

## Steps to reproduce the problem

Create an empty ruby file, add the following piece of code to it:
```ruby
a = {
  b: "{\"[\\\"*\\\"]\""
}
```

run:
```bash
rubocop -a [name-of-the-file]
```

check the file again.

## RuboCop version

```bash
$ rubocop -V
1.37.1 (using Parser 3.1.2.1, rubocop-ast 1.23.0, running on ruby 2.7.6) [x86_64-linux]
```

**Comment by konieczkow** [comment]
Had a look at the source code and it seems to me that the problem is somewhere here: 
https://github.com/rubocop/rubocop/blob/e2f5f20b7fe35b360795b63c1bd22bb49a93091b/lib/rubocop/cop/util.rb#L125

Switching the order of the `gsub`s fixes my issue:

```ruby
original_string = "{\"[\\\"*\\\"]\"".freeze

irb(main):038:0> original_string.gsub('\"', '"').gsub('\\') { '\\\\' }
=> "{\"[\"*\"]\""
irb(main):039:0> original_string.gsub('\\') { '\\\\' }.gsub('\"', '"')
=> "{\"[\\\"*\\\"]\""
```


but not sure if it doesn't break anything else. Will clone the project and run the test suite to see.

**Comment by konieczkow** [comment]
Running the specs - this breaks a few examples in `spec/rubocop/cop/style/quoted_symbols_spec.rb`.

**Comment by tdeo** [comment]
@konieczkow if you want to open a PR for this it seems to me the issue is actually that `QuotedSymbols` calls `to_string_literal` not with the intended argument style. 

Indeed, `StringLiterals` through `StringLiteralCorrector` calls it with the result of `DStrNode.str_content`, which takes care of handling some escaping:

https://github.com/rubocop/rubocop/blob/7007fb10dd223c54d131878ff460dedb718b85b3/lib/rubocop/cop/correctors/string_literal_corrector.rb#L13-L15

While `QuotedSymbols` calls it directly with the source of the node which retains escape-backslashes as characters:

https://github.com/rubocop/rubocop/blob/7007fb10dd223c54d131878ff460dedb718b85b3/lib/rubocop/cop/style/quoted_symbols.rb#L72-L77

It seems that on top of your fix, changing those lines to using `eval` to replicate the behavior of `str_content` actually makes the tests pass for `Style/QuotedSymbols`:
```ruby
# Using eval to replicate the behavior of `#str_content`
str = if hash_colon_key?(node)
                  correct_quotes(eval(node.source))
                else
                  # strip leading `:`
                  ":#{correct_quotes(eval(node.source[1..]))}"
                end
```

But I'm not sure whether using `eval` here would be accepted - maybe the proper solution is to implement `str_content` on `SymNode` that behaves similarly as for `DstrNode`. Maybe a rubocop maintainer could point in the preferred direction for implementation?
